terrarium/bi_ci/bi_ci/execute_mypy_multi.py

```diff
@@ -13,52 +16,155 @@
+
+    async def get(self) -> Path:
+        return await self._queue.get()
+
+    async def put(self, value: Path) -> None:
+        await self._queue.put(value)
+
+
 def get_mypy_targets(pkg_dir: Path) -> list[str]:
-    try:
-        with open(pkg_dir / PYPROJECT_TOML) as fh:
-            meta = tomlkit.load(fh)
+    with open(pkg_dir / PYPROJECT_TOML) as fh:
+        meta = tomlkit.load(fh)
+        try:
             return meta["datalens"]["meta"]["mypy"]["targets"]  # type: ignore  # 2024-01-30 # TODO: Value of type "Item | Container" is not indexable  [index]
-    except NonExistentKey:
-        pass
+        except NonExistentKey:
+            print("No data in meta['datalens']['meta']['mypy']['targets']")
+        try:
+            # just run mypy only on application/lib code
+            # we don't run mypy on tests for backward compatability because there are a lot of mypy issues already
+            targets: list[str] = [
+                package["include"]  # type: ignore # 2024-03-08 # Invalid index type "str" for "Any | str"; expected type "SupportsIndex | slice"  [index]
+                for package in meta["tool"]["poetry"]["packages"]  # type: ignore  # 2024-03-08 # TODO: Value of type "Item | Container" is not indexable  [index]
+            ]
+            return targets
+        except NonExistentKey:
+            print("No data in meta['tools']['poetry']['packages']")
 
-    # fallback to the same name defaults
+    # fallback
     dirname = pkg_dir.name or ""
     return [dirname]
 
 
-def get_targets(root: Path) -> Iterable[str]:
+def get_python_packages(root: Path) -> Iterable[str]:
     for path in root.rglob(f"*{PYPROJECT_TOML}"):
         yield str(path.parent).replace(PYPROJECT_TOML, "")
 
 
-def main(root: Path, targets_file: Path = None) -> None:  # type: ignore  # 2024-01-30 # TODO: Incompatible default for argument "targets_file" (default has type "None", argument has type "Path")  [assignment]
+async def run_mypy(request: MypyRequest, cache_queue: CacheQueue) -> MypyResult:
+    mypy_cache_dir = await cache_queue.get()
+    mypy_cache_dir.mkdir(exist_ok=True)
+
+    command_args = ["mypy", f"--cache-dir={mypy_cache_dir}", *request.targets]
+    command = " ".join(command_args)
+    proc = await asyncio.create_subprocess_shell(
+        command,
+        stdout=asyncio.subprocess.PIPE,
+        stderr=asyncio.subprocess.PIPE,
+        cwd=str(request.pkg_dir),
+    )
+    stdout, stderr = await proc.communicate()
+    await cache_queue.put(mypy_cache_dir)
+    return MypyResult(
+        pkg_dir=request.pkg_dir,
+        command=command,
+        status=MypyStatus.SUCCESS if proc.returncode == 0 else MypyStatus.FAIL,
+        stdout=stdout,
+        stderr=stderr,
+    )
+
+
+def main(root: Path, targets_file: Path = None, processes: int = 1) -> None:  # type: ignore  # 2024-01-30 # TODO: Incompatible default for argument "targets_file" (default has type "None", argument has type "Path")  [assignment]
```

Comment:
mb targets_file: typing.Optional[Path] then?

Reply:
clize can't recognize type annotation "Optional"
https://github.com/datalens-tech/datalens-backend/pull/373/files#diff-0cbab913b86b0d1b95f6576f34980e396c322821c8cca7d4fbdaba43d7755f51L35
I'll move it a little higher

Reply:
let's just put it in type: ignore instead of default TODO